### PR TITLE
magento/magento2#12817: [Forwardport] Coupon code with canceled order.

### PR DESCRIPTION
**Pull request by p-bystritsky**
Forwardport for https://github.com/magento-engcom/magento2ce/pull/1095

### Description

The number of coupon code uses is decreased if an order, that previously used this coupon code, is canceled.

### Fixed Issues (if relevant)

1. magento/magento2#12817: Coupon code with canceled order

### Manual testing scenarios

1. Create a Cart Price Rule with Specific Coupon. Set Uses per Coupon to 1.
2. Create Order (1st Order).
3. Apply the Coupon to 1st Order.
4. Place the 1st Order.
5. Create Order (2nd Order).
6. Apply the Coupon to 2nd Order.
7. Message "Coupon code is not valid" appears.
8. Cancel 1st Order.
9. Apply the Coupon to 2nd Order.
10. Message "Your coupon was successfully applied." appears.
11. Create a Cart Price Rule with Specific Coupon. Set Uses per Customer to 1.
12. Repeat steps 2-10 with this Coupon (note than orders must be placed by logged customer). 

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by magento-engcom-team** [comment]
Hi @p-bystritsky. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me 2.3-develop instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation

**Comment by PascalBrouwers** [comment]
@slavvka Can you just approve this? Because it comes from 2.2.4 branch.

**Comment by davidverholen** [comment]
closed and reopened to trigger new checks

File: app/code/Magento/SalesRule/Model/Coupon/UpdateCouponUsages.php [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    }
+
+    /**
+     * Executes the current command.
+     *
+     * @param Order $subject
+     * @param bool $increment
+     * @return Order
+     */
+    public function execute(Order $subject, bool $increment)
```
Comment (davidverholen) [inline app/code/Magento/SalesRule/Model/Coupon/UpdateCouponUsages.php]: method should have explicit return type
```suggestion
    public function execute(Order $subject, bool $increment): Order
```

File: app/code/Magento/SalesRule/Model/Coupon/UpdateCouponUsages.php [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+    }
+
+    /**
+     * Update the number of rule usages per customer.
+     *
+     * @param bool $increment
+     * @param int $ruleId
+     * @param int $customerId
+     */
+    private function updateCustomerRuleUsages(bool $increment, int $ruleId, int $customerId)
```
Comment (davidverholen) [inline app/code/Magento/SalesRule/Model/Coupon/UpdateCouponUsages.php]:
```suggestion
    private function updateCustomerRuleUsages(bool $increment, int $ruleId, int $customerId): void
```

File: app/code/Magento/SalesRule/Model/Coupon/UpdateCouponUsages.php [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+    }
+
+    /**
+     * Update the number of coupon usages.
+     *
+     * @param Order $subject
+     * @param bool $increment
+     * @param int $customerId
+     */
+    private function updateCouponUsages(Order $subject, bool $increment, int $customerId)
```
Comment (davidverholen) [inline app/code/Magento/SalesRule/Model/Coupon/UpdateCouponUsages.php]:
```suggestion
    private function updateCouponUsages(Order $subject, bool $increment, int $customerId): void
```

File: app/code/Magento/SalesRule/Model/ResourceModel/Coupon/Usage.php [hunk 4]
```diff
@@ -27,9 +27,10 @@ protected function _construct()
      *
      * @param int $customerId
      * @param mixed $couponId
+     * @param bool $increment
      * @return void
      */
-    public function updateCustomerCouponTimesUsed($customerId, $couponId)
+    public function updateCustomerCouponTimesUsed($customerId, $couponId, $increment = true)
```
Comment (davidverholen) [inline app/code/Magento/SalesRule/Model/ResourceModel/Coupon/Usage.php]:
```suggestion
    public function updateCustomerCouponTimesUsed($customerId, $couponId, $increment = true): void
```

**Review by davidverholen — CHANGES_REQUESTED** [review]

**Comment by p-bystritsky** [comment]
@davidverholen our suggestions committed.

**Review by davidverholen — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @davidverholen, thank you for the review.
[ENGCOM-5171](https://jira.corp.magento.com/browse/ENGCOM-5171) has been created to process this Pull Request

**Comment by VasylShvorak** [comment]
:heavy_check_mark: QA passed

**Comment by davidverholen** [comment]
@p-bystritsky why did you move it back to changes requested?

**Comment by p-bystritsky** [comment]
> @p-bystritsky why did you move it back to changes requested?

@davidverholen we had some non-public issues, but now it seems like they are fixed.

**Comment by engcom-Delta** [comment]
:heavy_check_mark: QA passed

**Pull request merged**